README.md

```diff
@@ -1,145 +1,308 @@
 
-cloud_sql_proxy authentication can be configured in a few different ways. Those listed higher on the list will override options lower on the list:
+* `-max_connections`
+
+If provided, the maximum number of connections to establish before refusing new
+connections. Defaults to 0 (no limit). Note: requests made through the
+SQL Admin API are rate limited to 2x the number of connections per minute.
+
+### Credentials and Authentication Flags
+
+Authentication can be configured in a few different ways. Those listed higher on
+the list will override options lower on the list:
 
 1. `credential_file` flag
 2. `token` flag
-3. Service account key at path stored in `GOOGLE_APPLICATION_CREDENTIALS`
-4. gcloud _user_ credentials (set from `gcloud auth login`)
-5. Default Application Credentials via goauth:
-
-   1. `GOOGLE_APPLICATION_CREDENTIALS` (again)
-   2. gcloud _application default_ credentials (set from ` gcloud auth application-default login`)
-   3. appengine.AccessToken (for App Engine Go < =1.9)
-   4. GCE/GAE metadata credentials
-
-When the proxy authenticates under the default service account of the
-Compute Engine VM it is running on the VM must have at least the
-sqlservice.admin API scope ("https://www.googleapis.com/auth/sqlservice.admin")
-and the associated project must have the SQL Admin API
-enabled.  The default service account must also have at least WRITER/EDITOR
-priviledges to any projects of target SQL instances.
-
-Specifying the `-credential_file` flag allows use of the proxy outside of
-Google's cloud. Simply [create a new service
+3. [Application Default Credentials](https://cloud.google.com/docs/authentication/production)
+
+Any account connecting to a Cloud SQL database will need one of the following
+IAM roles:
+
+- Cloud SQL Client (preferred)
+- Cloud SQL Editor
+- Cloud SQL Admin
+
+Or you can manually assign the following IAM permissions:
+
+- cloudsql.instances.connect
+- cloudsql.instances.get
+
+See [Roles and Permissions in Cloud
+SQL](https://cloud.google.com/sql/docs/mysql/roles-and-permissions) for details.
+
+When the proxy authenticates under the default service account of the Compute
+Engine VM it is running on the VM must have at least the sqlservice.admin API
+scope ("https://www.googleapis.com/auth/sqlservice.admin") and the associated
+project must have the SQL Admin API enabled. The default service account must
+also have at least WRITER/EDITOR priviledges to any projects of target SQL
+instances.
+
+* `-credential_file`
+
+Allows use of the proxy outside of Google's cloud. Simply [create a new service
 account](https://cloud.google.com/sql/docs/mysql/sql-proxy#create-service-account),
 download the associated JSON file, and set `-credential_file` to the path of the
-JSON file. You can also set the GOOGLE_APPLICATION_CREDENTIALS environment variable
-instead of passing this flag.
+JSON file. You can also set the `GOOGLE_APPLICATION_CREDENTIALS` environment
+variable instead of passing this flag.
```

Comment:
nit: This section feels very verbose. Can we try to make more concise? Maybe:

"Specifies the path to a JSON [service account](https://cloud.google.com/iam/docs/service-accounts) key the proxy uses to authorize or authenticate connections."